### PR TITLE
Debugger: Add Saved Addresses tab widget for bookmarking memory addresses

### PR DESCRIPTION
**Pull request by Daniel-McCarthy**
### Description of Changes
Adds the ability to save memory addresses to a newly added `Saved Addresses` tab widget in the debugger.
Works as a way to bookmark memory addresses and allow jumping back to saved addresses.
PCSX2-SavedAddresses

Adds the ability to add memory addresses as a Saved Addresses from the Memory Search results list context menu and `Memory View` context menu. Also allows exporting and importing the Saved Addresses in CSV format so that users can save/restore their addresses.

### Additional changes
- Updates double clicking Memory Search results to change the current tab to the `Memory View`. Previously it would change the address but stay on the same widget (thus not showing the address to the user).
- Replaces uses of the number `256` as a literal in place of `Qt::UserRole` to make it more clear what it's purpose is/where the number stems from.
- Const changes to address warnings in otherwise unrelated code in `CpuWidget.cpp`.

### Rationale behind Changes
Helps making keeping track of memory addresses/search results easier by allowing the user to give labels/descriptions and immediately jump back to that address in the `Memory View` or `Disassembly View`.

This makes it easier to juggle between different memory addresses and remembering details about each, and being able to quickly compare different areas of memory.

### Suggested Testing Steps
- Verify adding a new Saved Address from: The `Saved Addresses` context menu; memory search result context menu; and `Memory View` context menu.
- Exporting/Importing CSV of saved addresses adds the expected addresses.
- Going to disassembly and memory views is working as expected.
- Double clicking a memory search result while on a non-`Memory View` tab takes you to the `Memory View` tab.

### CODE REVIEW

**Comment by stenzek** [comment]
RTTI is disabled in our builds, but `qobject_cast` should still work.

**Review by Mrlinkwii — APPROVED** [review]
image
works well

**Review by kamfretoz — APPROVED** [review]
LGTM

Screenshot_20240101_180756

**Review by JordanTheToaster — APPROVED** [review]

File: pcsx2-qt/CMakeLists.txt [hunk 1]
```diff
@@ -167,6 +167,8 @@ target_sources(pcsx2-qt PRIVATE
 	Debugger/Models/ThreadModel.h
 	Debugger/Models/StackModel.cpp
 	Debugger/Models/StackModel.h
+    Debugger/Models/SavedAddressesModel.cpp
```
Comment (RedDevilus) [inline pcsx2-qt/CMakeLists.txt:170]: Tab, instead of 4 spaces

File: pcsx2-qt/CMakeLists.txt [hunk 2]
```diff
@@ -167,6 +167,8 @@ target_sources(pcsx2-qt PRIVATE
...
+    Debugger/Models/SavedAddressesModel.h
```
Comment (RedDevilus) [inline pcsx2-qt/CMakeLists.txt:171]: You are using 4 spaces instead of a tab.

File: pcsx2-qt/Debugger/CpuWidget.cpp [hunk 3]
```diff
@@ -39,13 +40,15 @@ CpuWidget::CpuWidget(QWidget* parent, DebugInterface& cpu)
 	, m_bpModel(cpu)
 	, m_threadModel(cpu)
 	, m_stackModel(cpu)
+    , m_savedAddressesModel(cpu)
```
Comment (RedDevilus) [inline pcsx2-qt/Debugger/CpuWidget.cpp:43]: You are using 4 spaces instead of a tab.

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2023  PCSX2 Dev Team
```
Comment (RedDevilus) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp:2]: Can also be 2024 now.

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2023  PCSX2 Dev Team
```
Comment (RedDevilus) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.h:2]: Can be changed to 2002-2024

**Review by RedDevilus — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

**Review by kamfretoz — APPROVED** [review]

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+ *  of the GNU Lesser General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  PCSX2 is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with PCSX2.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
```
Comment (stenzek) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp:14]: Could you please swap these over to the short-form SPDX headers? i.e.

```cpp
// SPDX-FileCopyrightText: 2002-2024 PCSX2 Dev Team
// SPDX-License-Identifier: LGPL-3.0+
```

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+			case HeaderColumns::ADDRESS:
+				return savedAddress.address;
+			case HeaderColumns::LABEL:
+				return savedAddress.label;
+			case HeaderColumns::DESCRIPTION:
+				return savedAddress.description;
+		}
+	}
+	return QVariant();
+
```
Comment (stenzek) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.cpp:59]: Extra whitespace

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	Qt::ItemFlags flags(const QModelIndex& index) const override;
+	void addRow();
+	void addRow(SavedAddress addresstoSave);
+	bool removeRows(int row, int count, const QModelIndex& parent = QModelIndex()) override;
+	bool setData(const QModelIndex& index, const QVariant& value, int role) override;
+
+private:
+	DebugInterface& m_cpu;
+	std::vector<SavedAddress> m_savedAddresses;
+
```
Comment (stenzek) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.h:64]: Extra whitespace

File: pcsx2-qt/Debugger/Models/SavedAddressesModel.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/*  PCSX2 - PS2 Emulator for PCs
```
Comment (stenzek) [inline pcsx2-qt/Debugger/Models/SavedAddressesModel.h:1]: Ditto with above

**Review by stenzek — COMMENTED** [review]
Looks good otherwise.

I know we were discussing whether to serialize this information in a JSON file, I'm guessing that is out of scope for this PR? Just wanted to check before merging.

**Comment by Daniel-McCarthy** [comment]
> I know we were discussing whether to serialize this information in a JSON file, I'm guessing that is out of scope for this PR? Just wanted to check before merging.

Yeah, I was thinking to merge this in and then will add the serialization/debugger settings as a followup. Thanks for checking 👍

**Comment by stenzek** [comment]
All good! Thanks for the PR :)

**Pull request merged**